cluster-logging: Add LokiStack tokenized auth proposal (#1503)
openshift/enhancements: enhancements/cluster-logging/loki-tokenized-auth-enablement.md

## Motivation

As previously accomplished by other Red Hat and third-party operators (e.g. OADP Operator, External DNS Operator, AWS EFS CSI Driver Operator) the following proposal seeks to use the CCO's CredentialsRequest API in addition to the OpenShift Console OLM capabilities to detect STS-enabled clusters and manage the whole operator lifecycle.

### User Stories

* As a LokiStack administrator I want to off-load STS configuration to CCO so that I do not need to manually manage STS credentials configuration on AWS and in turn a custom LokiStack S3 object storage secret.
* As a LokiStack administrator I want to off-load Azure WIF configuration to CCO so that I do not need to manually manage WIF credentials configuration on Azure and in turn a custom LokiStack Azure object storage secret.
* As a LokiStack administrator I want to off-load GCP WIF configuration to CCO so that I do not need to manually manage WIF credentials configuration on GCP and in turn a custom LokiStack GCS object storage secret.

> 

Reviewer: Are we supporting only user stories that use CCO?

Reviewer 2: > Are we supporting only user stories that use CCO?

This verbiage implies the need to use the CCO utility but, aside from this proposal, it should be possible to use any flow that provides the necessary token.  My understanding is CCO simplifies the setup which results in a secret with the needed information.

Author: IMHO we should limit our support vector here because an operator bundle with the `features.operators.openshift.io/token-auth-{aws,azure,gcp}` annotation will only work in companion with the CCO workflow on OpenShift. Although we support upstream a setup w/o CCO, I suggest to support only the CCO workflow for OCP subscriptions.

Allow LokiStack administrators to manage LokiStack Cloud Provider related configuration (i.e. object storage service access) like with any other Red Hat Operator enabled for tokenized authentication.

Allow OpenShift administrators to pre-provision Cloud Provider IAM resources for the layered product Red Hat OpenShift Logging and specifically the Loki Operator before installation.

Allow Hosted Control Planes (HCP) and Red Hat OpenShift on AWS (ROSA) cluster administrators to to pre-provision Hosted Provider IAM resources for the layered product Red Hat OpenShift Logging and specifically the Loki Operator before installation.

### Non-Goals

Automating pre-provisioned Cloud Providers' IAM resources so that users are left only with the Loki Operator and the LokiStack resource installation only.

## Proposal

The Loki Operator supports per Cloud Provider the following annotations in it's respectives ClusterServiceVersion per bundle (community-openshift, openshift):

- For AWS STS: `features.operators.openshift.io/token-auth-aws: true`
- For Azure WIF: `features.operators.openshift.io/token-auth-azure: true`
- For GCP WIF: `features.operators.openshift.io/token-auth-gcp: true`

> These will enable to operator installation to require specific IAM resource identifiers (e.g. role, client id, tenant id, etc.) which will be populated to the subscription and in turn exposed in the operator container environment.

Reviewer 2: Some subtle grammar changes, but I would shorten this since it is long:

 These enable the operator installation to require specific IAM resource identifiers (e.g. role, client id, tenant id, etc.).  The operator subscription is populated with the identifiers and exposed in the operator container environment.

### Workflow Description

> The workflow implemented by this proposal follows exact the flow presented in the [Tokenized Authentication Enabledment for Red Hat Operators on Cloud Providers](tokenized-auth-enablement-workflow) enhancement proposal.

Reviewer 2: ... exactly follows..

you could also just say "follows"

For the LokiStack administrator using the OpenShift Console - Operator installation under the proposed system:

1. user goes to console and starts operator install
2. console detects token-based (STS) cluster, and token-based-supporting operator

> 3. console prompts user to create roles+supply ARN-type string

Reviewer 2: roles and supply

4. console creates subscription with ARN-type string embedded as a spec.config.env
5. operator deployment is created with ARN-type string in env
6. operator creates `CredentialsRequest` including the ARN-type string
7. Cloud Credential Operator populates Secret based on `CredentialsRequest`

> 8. operator loads Secret, makes cloud service requests

Reviewer 2: Secret and makes

> For the LokiStack admnistator using the OpenShift CLI - Operator installation under the proposed system:

Reviewer 2: ..administrator..

- For AWS:
```yaml
kind: Subscription
metadata:
 name: ...
spec:
  config:
    env:
    - name: ROLEARN
      value: "<role ARN >"
```

- For Azure:
```yaml
kind: Subscription
metadata:
 name: ...
spec:
  config:
    env:
    - name: CLIENT_ID
      value: "<azure client id>"
    - name: TENANT_ID
      value: "<azure tenant id>"
    - name: SUBSCRIPTION_ID

>       value: "<azure subscription id>""

Reviewer: ```suggestion
      value: "<azure subscription id>"
```

```

- For GCP: TBD along 4.15.

For the Loki Operator author team:
- Add code to create a `CredentialRequest` per `LokiStack` resource.
- Add IAM-idetifier-type fields to the `CredentialRequest` resources.
- Add eventing to report status on a CR to indicate lacking STS credentials for fully operational deploy or update.

#### AWS STS Support

> The Loki Operator will use the `ROLEARN` environment variable and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The `ROLEARN` value is placed in the `spec.providerSpec.stsIAMRoleARN` field. The `CredentialsRequest` resource references the main LokiStack serviceaccount name that is the same as the LokiStack resource name.

Reviewer: ```suggestion
The Loki Operator will use the `ROLEARN` environment variable and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The `ROLEARN` value is placed in the `spec.providerSpec.stsIAMRoleARN` field. The `CredentialsRequest` resource references the main LokiStack service account name which is the same as the LokiStack resource name. The LokiStack administrator is required to provide an AWS Region via the AWS object storage secret.
```

```yaml
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: logging-loki-aws-creds
  namespace: openshift-cloud-credential-operator
spec:
  cloudTokenPath: /var/run/secrets/openshift/serviceaccount/token
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - s3:ListBucket
      - s3:PutObject
      - s3:GetObject
      - s3:DeleteObject
      effect: Allow
      resource: arn:aws:s3:*:*:*
    stsIAMRoleARN: arn:aws:iam::${AWS_ACCOUNT_ID}:role/openshift-logging-loki-operator-controller-manager
  secretRef:
    name: logging-loki-aws-creds
    namespace: openshift-logging
  serviceAccountNames:
  - logging-loki
```

#### Azure WIF Support

> The Loki Operator will use the provided environment variables (`CLIENT_ID`, `TENANT_ID`, `SUBSCRIPTION_ID`) and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The values will be used in the `CredentialsRequest` fields `spec.providerSpec.{AzureClientID,AzureTenantID,AzureSubscriptionID}` respectively. The LokiStack administrator is required to provider an Azure Region via the Azure object storage secret.

Reviewer: There is a discrepancy between the fields mentioned here and in the [design doc](https://github.com/grafana/loki/pull/11060/files#diff-27b5aa2c87cf76afa33759f0ba55f749b2a54e1d4b1b6c7bc69a2cf1cbfc1c24R69-R76). In the design doc we mention `AzureClientID,AzureTenantID` however CCO seems to care about some other fields `AzureSubscriptionID, AzureRegion`.

Author: That's correct and well known because my investigation in the Loki code base did not uncover any usage of the azure subscription id and/or region. I am leaving this open here for the research task in LOG-4545. WDYT?

> The Loki Operator will use the provided environment variables (`CLIENT_ID`, `TENANT_ID`, `SUBSCRIPTION_ID`) and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The values will be used in the `CredentialsRequest` fields `spec.providerSpec.{AzureClientID,AzureTenantID,AzureSubscriptionID}` respectively. The LokiStack administrator is required to provider an Azure Region via the Azure object storage secret.

Reviewer 2: ...is required to provide an Azure...

```yaml
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: logging-loki-azure-creds
  namespace: openshift-cloud-credential-operator
spec:
  cloudTokenPath: /var/run/secrets/openshift/serviceaccount/token
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AzureProviderSpec
    AzureClientID: clientid
    AzureTenantID: tenantid
    AzureRegion: centralus
    AzureSubscriptionID: subscriptionid
  secretRef:
    name: logging-loki-azure-creds
    namespace: openshift-logging
  serviceAccountNames:
  - logging-loki
```

#### GCP WIF Support

TBD along 4.15.

### API Extensions

No further API Extensions needed.

### Implementation Details/Notes/Constraints [optional]

> Based on the reconciled `CredentialRequest` resource the Loki Operator will populated the credential secrets as following plus additional configuration required by Loki to operate with tokenized authentication.

Reviewer 2: ...secrets as follows...

> Based on the reconciled `CredentialRequest` resource the Loki Operator will populated the credential secrets as follows plus additional configuration required by Loki to operate with tokenized authentication.

Reviewer: Rephrased feel free to adopt it or re-write it once more
```suggestion
The Loki Operator will generate the Loki configuration needed to operate with tokenized authentication based on the credential secret created by the reconciliation of the `CredentialRequest` resource. This Loki configuration will change depending on the Cloud Provider as follows.
```

__Note__: The LokiStack components requiring access to object storage are: `ingester`, `querier`, `index-gateway`, `compactor` and `ruler`.

#### AWS STS Support

For AWS assuming the following credentials secret provided by CCO:

```yaml
apiVersion: v1
stringData:
  credentials: |-
    [default]
    role_arn = <role ARN>
    web_identity_token_file = <path to mounted service account token such /var/run/secrets/openshift/serviceaccount/token>
kind: Secret
metadata:
  name: logging-loki-aws-credentials
  namespace: openshift-logging
type: Opaque
```

> each Loki container will gain two new evironment variables:

Reviewer: ```suggestion
Each Loki container will gain two new environment variables:
```

```yaml
containers:
  - args:
    - -target=ingester
    - -config.file=/etc/loki/config/config.yaml
    - -runtime-config.file=/etc/loki/config/runtime-config.yaml
    - -config.expand-env=true
    env:
    - name: AWS_ROLE_ARN
      value: arn:aws:iam::269733383066:role/openshift-logging-loki-operator-controller-manager
    - name: AWS_WEB_IDENTITY_TOKEN_FILE
      value: /var/run/secrets/openshift/serviceaccount/token
```

> in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provid an `s3_storage_config` as below:

Reviewer: ```suggestion
in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:
```

> in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:

Reviewer: ```suggestion
In addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:
```

```yaml
common:
  storage:
    s3:
      s3: "s3://<region>/<bucketnames>"
      s3forcepathstyle: false
```

__Note:__ The above configuration is a workaround as suggested by the [grafana/loki#4269](grafana-loki-issue-4269) issue.

#### Azure WIF Support

For Azure assuming the following credentials secret provided by CCO:

```yaml
apiVersion: v1
stringData:
  azure_client_id: <client id>
  azure_federated_token_file: <path to mounted service account token such as /var/run/secrets/openshift/serviceaccount/token>
  azure_region: <region>
  azure_subscription_id: <subscription id>
  azure_tenant_id: <tenant id>
kind: Secret
metadata:
  name: logging-loki-aws-credentials
  namespace: openshift-logging
type: Opaque
```

> each Loki container will gain the following evironment variables:

Reviewer: ```suggestion
Each Loki container will gain the following environment variables:
```

```yaml
containers:
  - args:
    - -target=ingester
    - -config.file=/etc/loki/config/config.yaml
    - -runtime-config.file=/etc/loki/config/runtime-config.yaml
    - -config.expand-env=true
    env:
    - name: AZURE_CLIENT_ID
      value: <client id>
    - name: AZURE_TENANT_ID
      value: <tenant id>
    - name: AZURE_REGION
      value: <azure region>
    - name: AZURE_SUBSCRIPTION_ID
      value: <subscription id>
    - name: AZURE_FEDERATED_TOKEN_FILE
      value: /var/run/secrets/openshift/serviceaccount/token
```

> __Note:__ The Azure Region value is populated from the azure object storage secret referenced by the `lokistack.spec.storage.secret.name`.

Reviewer: I think we can remove this, this seems to be from one of the first iterations of this document, region seems to also be in the secret created by CCO so we can use that.

#### GCP WIF Support

TBD along 4.15.

### Risks and Mitigations

### Drawbacks

## Design Details

We have one PoC item:
How Loki Operator can be changed to work with the modified CCO for AWS STS: https://github.com/grafana/loki/compare/main...periklis:loki:aws-sts-auth-spike

### Test Plan

Consider the following in developing a test plan for this enhancement:
- TBD

### Graduation Criteria

#### Dev Preview -> Tech Preview

None.

#### Tech Preview -> GA

None.

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

### Version Skew Strategy

### Operational Aspects of API Extensions

#### Failure Modes

#### Support Procedures

## Implementation History

Major milestones in the life cycle of a proposal should be tracked in `Implementation
History`.

> 

Reviewer: ```suggestion
- AWS STS Support https://github.com/grafana/loki/pull/11481
```

```yaml
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: logging-loki-azure-creds
  namespace: openshift-logging
spec:
  cloudTokenPath: /var/run/secrets/storage/serviceaccount/token
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AzureProviderSpec
    AzureClientID: clientid
    AzureTenantID: tenantid
    AzureRegion: centralus
    AzureSubscriptionID: subscriptionid
  secretRef:
    name: logging-loki-azure-creds
    namespace: openshift-logging
  serviceAccountNames:
  - logging-loki
  - logging-loki-ruler
```

#### GCP WIF Support

TBD along 4.16.

### API Extensions

> No further API Extensions needed.

Reviewer: We should mention `CredentialMode` that was added recently

apiVersion: v1
stringData:
  azure_client_id: <client id>
  azure_federated_token_file: <path to mounted service account token such as /var/run/secrets/openshift/serviceaccount/token>
  azure_region: <region>
  azure_subscription_id: <subscription id>
  azure_tenant_id: <tenant id>
kind: Secret
metadata:
  name: logging-loki-aws-credentials
  namespace: openshift-logging
type: Opaque
```

Each Loki container will gain the following environment variables:

```yaml
containers:
  - args:
    - -target=ingester
    - -config.file=/etc/loki/config/config.yaml
    - -runtime-config.file=/etc/loki/config/runtime-config.yaml
    - -config.expand-env=true
    env:
    - name: AZURE_CLIENT_ID
      value: <client id>
    - name: AZURE_TENANT_ID
      value: <tenant id>
    - name: AZURE_SUBSCRIPTION_ID
      value: <subscription id>
    - name: AZURE_FEDERATED_TOKEN_FILE
      value: /var/run/secrets/storage/serviceaccount/token
```

__Note:__ The Azure Region value is populated from the azure object storage secret referenced by the `lokistack.spec.storage.secret.name`.

#### GCP WIF Support

TBD along 4.16.

### Risks and Mitigations

### Drawbacks

## Design Details

We have one PoC item:

> How Loki Operator can be changed to work with the modified CCO for AWS STS: https://github.com/grafana/loki/compare/main...periklis:loki:aws-sts-auth-spike

Reviewer: This link should be updated to https://github.com/grafana/loki/pull/11060

metadata:
  name: logging-loki
  namespace: openshift-logging
spec:
  cloudTokenPath: /var/run/secrets/storage/serviceaccount/token
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AzureProviderSpec
    AzureClientID: clientid
    AzureTenantID: tenantid
    AzureRegion: centralus
    AzureSubscriptionID: subscriptionid
  secretRef:
    name: logging-loki-managed-credentials
    namespace: openshift-logging
  serviceAccountNames:
  - logging-loki
  - logging-loki-ruler
```

#### GCP WIF Support

GCP WIF support will be supported only in the form of the upstream support as per CCO missing respectable bits.

### API Extensions

The present proposal introduces a new optional field in the `ObjectStorageSecretSpec` namely `CredentialsMode`. If the user does not provide a value the `CredentialsMode` is automatically detected either from the provided secret fields or the operator environment variables (latter applies only in OpenShift cluters.). The selected or detected `CredentialsMode` is populated in addition in the `status.storage.credentialMode` field.

```go
/ CredentialMode represents the type of authentication used for accessing the object storage.
//

> // +kubebuilder:validation:Enum=static;token;managed

Reviewer: ```suggestion
// +kubebuilder:validation:Enum=static;token;token-cco
```

type CredentialMode string

const (
    // CredentialModeStatic represents the usage of static, long-lived credentials stored in a Secret.
    // This is the default authentication mode and available for all supported object storage types.
    CredentialModeStatic CredentialMode = "static"
    // CredentialModeToken represents the usage of short-lived tokens retrieved from a credential source.
    // In this mode the static configuration does not contain credentials needed for the object storage.
    // Instead, they are generated during runtime using a service, which allows for shorter-lived credentials and
    // much more granular control. This authentication mode is not supported for all object storage types.
    CredentialModeToken CredentialMode = "token"

>     // CredentialModeManaged represents the usage of short-lived tokens retrieved from a credential source.
>     // This mode is similar to CredentialModeToken,but instead of having a user-configured credential source,
>     // it is configured by the environment, for example the Cloud Credential Operator in OpenShift.
>     // This mode is only supported for certain object storage types in certain runtime environments.
>     CredentialModeManaged CredentialMode = "managed"

Reviewer: ```suggestion
    // CredentialModeTokenCCO represents the usage of short-lived tokens retrieved from a credential source.
    // This mode is similar to CredentialModeToken, but instead of having a user-configured credential source,
    // it is configured by the environment and the operator relies on the Cloud Credential Operator to provide
    // a secret. This mode is only supported for certain object storage types in certain runtime environments.
    CredentialModeTokenCCO CredentialMode = "token-cco"
```

)

// ObjectStorageSecretSpec is a secret reference containing name only, no namespace.
type ObjectStorageSecretSpec struct {
...
    // CredentialMode can be used to set the desired credential mode for authenticating with the object storage.
    // If this is not set, then the operator tries to infer the credential mode from the provided secret and its
    // own configuration.
    //
    // +optional
    // +kubebuilder:validation:Optional
    CredentialMode CredentialMode `json:"credentialMode,omitempty"`
}
```

> The purpose of the `CredentialMode` is to override the detected credentials type from object storage secrets or the operator environment variables. Latter is only supported on AWS-STS/Azure-WIF managed OpenShift clusters, where the operator is using CredetialMode `managed` by default. However, the user might want to use `static` to store logs for example to Minio on the same cluster instead to AWS S3.

Reviewer: ```suggestion
The purpose of the `CredentialMode` is to override the detected credentials type from object storage secrets or the operator environment variables. Latter is only supported on AWS-STS/Azure-WIF managed OpenShift clusters, where the operator is using CredetialMode `token-cco` by default. However, the user might want to use `static` to store logs for example to Minio on the same cluster instead to AWS S3.
```
